File: ports/go/generator.go
package main

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

//go:embed template.html
var templateHTML string

// Payload is the data structure injected into the template.
type Payload struct {
	Dashboard         interface{} `json:"dashboard"`
	Drilldown         interface{} `json:"drilldown"`
	Analysis          interface{} `json:"analysis"`
	Recommendations   interface{} `json:"recommendations"`
	WorkDays          interface{} `json:"work_days"`
	Models            interface{} `json:"models"`
	Subagents         interface{} `json:"subagents"`
	Branches          interface{} `json:"branches"`
	ContextEfficiency interface{} `json:"context_efficiency"`
	Versions          interface{} `json:"versions"`
	Skills            interface{} `json:"skills"`
	SlashCommands     interface{} `json:"slash_commands"`
	Config            interface{} `json:"config"`
}

// GenerateHTML generates the final HTML dashboard.
func GenerateHTML(data *ParseResult, recommendations RecommendationResult) string {
	payload := Payload{
		Dashboard:         data.Dashboard,
		Drilldown:         data.Drilldown,
		Analysis:          data.Analysis,
		Recommendations:   recommendations,
		WorkDays:          data.WorkDays,
		Models:            data.Models,
		Subagents:         data.Subagents,
		Branches:          data.Branches,
		ContextEfficiency: data.ContextEfficiency,
		Versions:          data.Versions,
		Skills:            data.Skills,
		SlashCommands:     data.SlashCommands,
		Config:            data.Config,
	}

	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		payloadJSON = []byte("{}")
	}

	html := strings.Replace(templateHTML, "__DATA_PLACEHOLDER__", string(payloadJSON), 1)
	return html
}

// WriteReport writes the HTML report to disk.
func WriteReport(html string, outputPath string) (string, error) {
	if outputPath == "" {
		cwd, err := os.Getwd()
		if err != nil {
			return "", err
		}
		outputDir := filepath.Join(cwd, "output")
		if err := os.MkdirAll(outputDir, 0o755); err != nil {
			return "", err
		}
		timestamp := time.Now().Format("20060102-150405")
		outputPath = filepath.Join(outputDir, fmt.Sprintf("claude-analytics-%s.html", timestamp))
		ensureGitignore(outputDir)
	}

	dir := filepath.Dir(outputPath)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}

	if err := os.WriteFile(outputPath, []byte(html), 0o644); err != nil {
		return "", err
	}

	return outputPath, nil
}

// ReadLastRun reads the last run timestamp from the .last-run marker file.
func ReadLastRun() string {
	cwd, err := os.Getwd()
	if err != nil {
		return ""
	}
	marker := filepath.Join(cwd, "output", ".last-run")
	data, err := os.ReadFile(marker)
	if err != nil {
		return ""
	}
	content := strings.TrimSpace(string(data))
	if len(content) >= 10 {
		return content[:10]
	}
	return ""
}

// SaveLastRun saves the current timestamp as the last run marker.
func SaveLastRun() {
	cwd, err := os.Getwd()
	if err != nil {
		return
	}
	outputDir := filepath.Join(cwd, "output")
	os.MkdirAll(outputDir, 0o755)
	marker := filepath.Join(outputDir, ".last-run")
	content := time.Now().Format("2006-01-02 15:04:05") + "\n"
	os.WriteFile(marker, []byte(content), 0o644)
}

func ensureGitignore(outputDir string) {
	// .gitignore inside output/
	innerGitignore := filepath.Join(outputDir, ".gitignore")
	if _, err := os.Stat(innerGitignore); os.IsNotExist(err) {
		content := "# Claude Analytics reports contain sensitive session data.\n" +
			"# Do NOT commit these files to version control.\n" +
			"*\n" +
			"!.gitignore\n"
		os.WriteFile(innerGitignore, []byte(content), 0o644)
	}

	// Also add 'output/' to the project root .gitignore
	rootGitignore := filepath.Join(filepath.Dir(outputDir), ".gitignore")
	data, err := os.ReadFile(rootGitignore)
	if err == nil {
		if !strings.Contains(string(data), "output/") {
			f, err := os.OpenFile(rootGitignore, os.O_APPEND|os.O_WRONLY, 0o644)
			if err == nil {
				f.WriteString("\n# Claude Analytics reports (contain sensitive data)\noutput/\n")
				f.Close()
			}
		}
	} else if os.IsNotExist(err) {
		os.WriteFile(rootGitignore, []byte("# Claude Analytics reports (contain sensitive data)\noutput/\n"), 0o644)
	}
}

File: ports/rust/src/generator.rs
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

use chrono::Local;

use crate::parser::{ParseResult, RecommendationsResult};

const TEMPLATE: &str = include_str!("template.html");

pub fn generate_html(data: &ParseResult, recommendations: &RecommendationsResult) -> String {
    // Build a combined payload matching the template's expected shape.
    // Since all our types derive Serialize, we construct a serde_json::Value
    // from the typed structs for template injection.
    let payload = serde_json::json!({
        "dashboard": data.dashboard,
        "drilldown": data.drilldown,
        "analysis": data.analysis,
        "recommendations": recommendations,
        "work_days": data.work_days,
        "models": data.models,
        "subagents": data.subagents,
        "branches": data.branches,
        "context_efficiency": data.context_efficiency,
        "versions": data.versions,
        "skills": data.skills,
        "slash_commands": data.slash_commands,
        "permission_modes": data.permission_modes,
        "config": data.config,
    });

    let payload_json = serde_json::to_string(&payload).unwrap_or_else(|_| "{}".to_string());
    TEMPLATE.replace("__DATA_PLACEHOLDER__", &payload_json)
}

pub fn write_report(html: &str, output_path: Option<&str>) -> Result<PathBuf, String> {
    let path = if let Some(p) = output_path {
        PathBuf::from(p)
    } else {
        let output_dir = std::env::current_dir()
            .map_err(|e| format!("Cannot get current dir: {}", e))?
            .join("output");
        fs::create_dir_all(&output_dir)
            .map_err(|e| format!("Cannot create output dir: {}", e))?;

        let timestamp = Local::now().format("%Y%m%d-%H%M%S").to_string();
        let filename = format!("claude-analytics-{}.html", timestamp);

        // Ensure output/ is gitignored
        ensure_gitignore(&output_dir);

        output_dir.join(filename)
    };

    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .map_err(|e| format!("Cannot create parent dir: {}", e))?;
    }

    fs::write(&path, html)
        .map_err(|e| format!("Cannot write report: {}", e))?;

    Ok(path)
}

pub fn read_last_run(output_dir: Option<&Path>) -> Option<String> {
    let dir = output_dir
        .map(|p| p.to_path_buf())
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_default().join("output"));
    let marker = dir.join(".last-run");
    if marker.exists() {
        if let Ok(content) = fs::read_to_string(&marker) {
            let content = content.trim();
            if content.len() >= 10 {
                return Some(content[..10].to_string());
            }
        }
    }
    None
}

pub fn save_last_run(output_dir: Option<&Path>) {
    let dir = output_dir
        .map(|p| p.to_path_buf())
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_default().join("output"));
    let _ = fs::create_dir_all(&dir);
    let marker = dir.join(".last-run");
    let content = Local::now().format("%Y-%m-%d %H:%M:%S\n").to_string();
    let _ = fs::write(&marker, content);
}

fn ensure_gitignore(output_dir: &Path) {
    // .gitignore inside output/
    let inner_gitignore = output_dir.join(".gitignore");
    if !inner_gitignore.exists() {
        let _ = fs::write(
            &inner_gitignore,
            "# Claude Analytics reports contain sensitive session data.\n\
             # Do NOT commit these files to version control.\n\
             *\n\
             !.gitignore\n",
        );
    }

    // Add 'output/' to project root .gitignore
    if let Some(parent) = output_dir.parent() {
        let root_gitignore = parent.join(".gitignore");
        if root_gitignore.exists() {
            if let Ok(content) = fs::read_to_string(&root_gitignore) {
                if !content.contains("output/") {
                    let _ = fs::OpenOptions::new()
                        .append(true)
                        .open(&root_gitignore)
                        .and_then(|mut f| {
                            writeln!(f, "\n# Claude Analytics reports (contain sensitive data)\noutput/")
                        });
                }
            }
        } else {
            let _ = fs::write(
                &root_gitignore,
                "# Claude Analytics reports (contain sensitive data)\noutput/\n",
            );
        }
    }
}
